src/core/raster/qgscolorrampshader.cpp

```diff
@@ -649,3 +649,135 @@ void QgsColorRampShader::setLegendSettings( QgsColorRampLegendNodeSettings *sett
+
+
+  if ( properties.contains( QStringLiteral( "color_ramp_shader_items_list" ) ) )
+  {
+    QVariant colorRampItemsVar = properties.value( QStringLiteral( "color_ramp_shader_items_list" ) );
+    if ( colorRampItemsVar.type() == QVariant::List )
+    {
+      QVariantList itemVariantList = colorRampItemsVar.toList();
+      for ( const QVariant &itemVar : std::as_const( itemVariantList ) )
+      {
+        QgsColorRampShader::ColorRampItem item;
+        if ( itemVar.type() != QVariant::Map )
+          continue;
+        QVariantMap itemVarMap = itemVar.toMap();
+        if ( !itemVarMap.contains( QStringLiteral( "label" ) ) || !itemVarMap.contains( QStringLiteral( "color" ) ) || !itemVarMap.contains( QStringLiteral( "value" ) ) )
+          continue;
+
+        item.label = itemVarMap.value( QStringLiteral( "label" ) ).toString();
+        item.color = QgsSymbolLayerUtils::decodeColor( itemVarMap.value( QStringLiteral( "color" ) ).toString() );
+        item.value = itemVarMap.value( QStringLiteral( "value" ) ).toDouble();
+
+        colorRampShader.mColorRampItemList.append( item );
+      }
+    }
+  }
+
+  if ( properties.contains( QStringLiteral( "color_ramp_shader_minimum_value" ) ) )
+    colorRampShader.setMinimumValue( properties.value( QStringLiteral( "color_ramp_shader_minimum_value" ) ).toDouble() );
+  else
+    colorRampShader.setMinimumValue( std::numeric_limits<double>::quiet_NaN() );
+
+  if ( properties.contains( QStringLiteral( "color_ramp_shader_maximum_value" ) ) )
+    colorRampShader.setMaximumValue( properties.value( QStringLiteral( "color_ramp_shader_maximum_value" ) ).toDouble() );
+  else
+    colorRampShader.setMaximumValue( std::numeric_limits<double>::quiet_NaN() );
+
+  if ( properties.contains( QStringLiteral( "color_ramp_shader_value_out_of_range" ) ) )
+    colorRampShader.setClip( properties.value( QStringLiteral( "color_ramp_shader_value_out_of_range" ) ).toInt() == 1 );
+  if ( properties.contains( QStringLiteral( "color_ramp_shader_label_precision" ) ) )
+    colorRampShader.setLabelPrecision( properties.value( QStringLiteral( "color_ramp_shader_label_precision" ) ).toInt() );
+
+  QDomDocument docLegendSettings;
+  if ( properties.contains( QStringLiteral( "color_ramp_shader_legend_settings" ) ) &&
+       docLegendSettings.setContent( properties.value( QStringLiteral( "color_ramp_shader_legend_settings" ) ).toString() ) )
+  {
+    QDomElement elemLegendProperties = docLegendSettings.firstChildElement( QStringLiteral( "legend_settings_properties" ) );
+    if ( !elemLegendProperties.isNull() )
+    {
+      std::unique_ptr<QgsColorRampLegendNodeSettings> legendSettings = std::make_unique<QgsColorRampLegendNodeSettings>();
+      legendSettings->readXml( elemLegendProperties, QgsReadWriteContext() );
+      colorRampShader.setLegendSettings( legendSettings.release() );
+    }
+  }
+
+  return colorRampShader;
+}
```

Comment:
Hmm I am not a big fan of this implementation to read/write color ramp shader from/to QVariantMap:
- source color ramp properties are mixed with shader properties - can we keep them separated? (like it is done with "color_ramp_shader_items_list")
- source color ramp properies could be probably loaded/saved with QgsSymbolLayerUtils::colorRampToVariant()  + QgsSymbolLayerUtils::loadColorRamp() - this would avoid the need to call create() method of various color ramp types directly
- legend settings are loaded/stored as embedded XML, which is suboptimal. And it looks like we do not need legend settings of the shader for interpolated line renderer anyway (though I understand you have added it for completeness)

I wonder if we could instead move these two functions as private methods to interpolated line symbol layer class... So it will be just implementation detail of the symbol layer, and not "official" API of QgsColorRampShader. By doing that, we could also afford to leave out the bits of information we don't need (like the mentioned legend settings). What do you think?

Reply:
done